### ISSUE TITLE
[ZScaler ZIA Web]: Lack of input sanitization leads to inconsistent behaviour when processing data

### ISSUE THREAD

**Issue by srilumpa** [elastic/integrations#10668]
### Integration Name

Zscaler Internet Access [zscaler_zia]

### Integration Version

3.0.2

### Agent Version

8.14

### Agent Output Type

elasticsearch

### Elasticsearch Version

8.14

### OS Version and Architecture

UbunTu 20.04 LTS

### Event Original

```
"{\"sourcetype\":\"zscalernss-web\",\"event\":{\"time\":\"Wed Jul 31 06:54:37 2024\",\"cloudname\":\"REDACTED\",\"host\":\"mail.163.com\",\"serverip\":\"1.2.3.4\",\"external_devid\":\"NA\",\"devicemodel\":\"NA\",\"action\":\"Allowed\",\"recordid\":\"999999999999\",\"reason\":\"Allowed\",\"threatseverity\":\"None\",\"tz\":\"GMT\",\"filesubtype\":\"jpeg\",\"upload_filesubtype\":\"None\",\"sha256\":\"None\",\"bamd5\":\"None\",\"filename\":\"Image.jpg\",\"upload_filename\":\"None\",\"filetype\":\"Jpeg Files\",\"devicename\":\"NA\",\"devicehostname\":\"NA\",\"deviceostype\":\"NA\",\"deviceosversion\":\"NA\",\"devicetype\":\"NA\",\"reqsize\":\"4575\",\"reqmethod\":\"GET\",\"refererurl\":\"mail.163.com/js6/read/readhtml.jsp?mid=REDACTED\",\"respsize\":\"245704\",\"respcode\":\"200\",\"reqversion\":\"1.1\",\"respversion\":\"1.1\",\"proto\":\"HTTPS\",\"company\":\"COMPANY\",\"dlpmd5\":\"None\",\"apprulelabel\":\"None\",\"dlprulename\":\"NA\",\"rulelabel\":\"None\",\"urlfilterrulelabel\":\"None\",\"cltip\":\"4.3.2.1\",\"cltintip\":\"4.3.2.1\",\"cltsourceport\":\"35710\",\"threatname\":\"None\",\"cltsslcipher\":\"TLS1_3_CK_AES_256_GCM_SHA384\",\"clttlsversion\":\"TLS1_3\",\"url\":\"mail.163.com/js6/s?func=mbox:getMessageData&mid=REDACTED\",\"useragent\":\"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0\",\"login\":\"user@company.com\",\"applayerprotocol\":\"N/A\",\"appclass\":\"Webmail\",\"appname\":\"163 Mail\",\"appriskscore\":\"3\",\"bandwidthclassname\":\"General Surfing\",\"bandwidthrulename\":\"No Bandwidth Control\",\"bwthrottle\":\"NO\",\"bypassedtime\":\"Thu Jan  1 00:00:00 1970\",\"bypassedtraffic\":\"0\",\"cltsslsessreuse\":\"NO\",\"cltpubip\":\"4.3.2.1\",\"cltsslfailcount\":\"0\",\"cltsslfailreason\":\"N/A\",\"contenttype\":\"image/jpeg\",\"datacentercity\":\"Xx\",\"datacentercountry\":\"XX\",\"datacenter\":\"XXXX\",\"day\":\"Wed\",\"day_of_month\":\"31\",\"dept\":\"Somewhere\",\"deviceappversion\":\"NA\",\"deviceowner\":\"NA\",\"df_hosthead\":\"None\",\"df_hostname\":\"None\",\"dlpdicthitcount\":\"None\",\"dlpdict\":\"None\",\"dlpeng\":\"None\",\"dlpidentifier\":\"0\",\"eedone\":\"No\",\"epochtime\":\"1722408877\",\"fileclass\":\"Images\",\"flow_type\":\"ZIA\",\"forward_gateway_ip\":\"0.0.0.0\",\"forward_gateway_name\":\"None\",\"forward_type\":\"None\",\"hour\":\"06\",\"is_sslexpiredca\":\"NO\",\"is_sslselfsigned\":\"NO\",\"is_ssluntrustedca\":\"No\",\"keyprotectiontype\":\"Software Protection\",\"location\":\"SOME_LOCATION\",\"malwarecategory\":\"None\",\"malwareclass\":\"None\",\"minute\":\"54\",\"mobappcategory\":\"None\",\"mobappname\":\"None\",\"mobdevtype\":\"None\",\"module\":\"Webmail\",\"month\":\"Jul\",\"month_of_year\":\"07\",\"nssserviceip\":\"192.168.0.100\",\"oapprulelabel\":\"None\",\"obwclassname\":\"General Surfing\",\"ocip\":\"428067240\",\"ocpubip\":\"428067240\",\"odevicehostname\":\"NA\",\"odevicename\":\"NA\",\"odeviceowner\":\"NA\",\"odlpdict\":\"None\",\"odlpeng\":\"None\",\"odlprulename\":\"None\",\"ofwd_gw_name\":\"None\",\"ologin\":\"2195246850\",\"ordr_rulename\":\"None\",\"ourlcat\":\"Webmail\",\"ourlfilterrulelabel\":\"None\",\"ozpa_app_seg_name\":\"None\",\"externalsslpolicyreason\":\"INSPECTED\",\"productversion\":\"6.2r.620_prod.84.400964_114\",\"rdr_rulename\":\"None\",\"refererhost\":\"mail.163.com\",\"reqheadersize\":\"4575\",\"reqdatasize\":\"0\",\"respheadersize\":\"353\",\"respdatasize\":\"245351\",\"riskscore\":\"0\",\"ruletype\":\"None\",\"second\":\"37\",\"srvcertchainvalpass\":\"PASS\",\"srvcertvalidationtype\":\"OV\",\"srvcertvalidityperiod\":\"MEDIUM\",\"srvsslcipher\":\"TLS1_3_CK_AES_256_GCM_SHA384\",\"serversslsessreuse\":\"NO\",\"srvocspresult\":\"None\",\"srvtlsversion\":\"TLS1_3\",\"srvwildcardcert\":\"YES\",\"ssldecrypted\":\"Yes\",\"throttlereqsize\":\"0\",\"throttlerespsize\":\"0\",\"totalsize\":\"250279\",\"trafficredirectmethod\":\"PAC\",\"unscannabletype\":\"None\",\"upload_doctypename\":\"None\",\"upload_fileclass\":\"None\",\"upload_filetype\":\"None\",\"urlcatmethod\":\"Database A\",\"urlsubcat\":\"Webmail\",\"urlsupercat\":\"Internet Communication\",\"urlclass\":\"Business Use\",\"useragentclass\":\"None\",\"useragenttoken\":\"Microsoft Chromium Edge (127.x)\",\"userlocationname\":\"Unknown\",\"year\":\"2024\",\"ztunnelversion\":\"None\",\"zpa_app_seg_name\":\"None\"}}"
```

### What did you do?

The above raw event (and many similar other events) is pushed directly into Elasticsearch to be ingested in the `logs-zscaler_zia.web-default` data stream and to be processed by the Zscaler ZIA Web pipeline. No specific configuration whatsoever on the Elasticsearch side. ZScaler NSS log forwarder was set as required in the integration documentation (https://www.elastic.co/docs/current/integrations/zscaler_zia#web-log).

### What did you see?

Ingested document was the following:

```json
{
  "zscaler_zia": {
    "web": {
      "referer": {
        "host": "mail.163.com"
      },
      "year": 2024,
      "alpn_protocol": "N/A",
      "day_of_month": 31,
      "second": 37,
      "user_location_name": "Unknown",
      "day": "Wed",
      "obfuscated": {
        "bendwidth": {
          "class_name": "General Surfing"
        },
        "client": {
          "public": {
            "ip": "428067240"
          },
          "ip": "428067240"
        },
        "login": "2195246850",
        "url": {
          "category": "Webmail"
        }
      },
      "ssl_decrypted": "Yes",
      "app": {
        "risk_score": "3",
        "name": "163 Mail",
        "class": "Webmail"
      },
      "module": "Webmail",
      "traffic_redirect_method": "PAC",
      "is_ssl_certificate_selfsigned": "NO",
      "forward_gateway": {
        "ip": "0.0.0.0"
      },
      "flow_type": "ZIA",
      "month": "Jul",
      "eedone": "No",
      "bypassed": {
        "time": "1970-01-01T00:00:00.000Z",
        "traffic": "0"
      },
      "key_protection_type": "Software Protection",
      "is_ssl_certificate_untrusted": "No",
      "request": {
        "payload": 0,
        "header_size": 4575
      },
      "server": {
        "cipher": "TLS1_3_CK_AES_256_GCM_SHA384",
        "certificate_validation_chain": "PASS",
        "certificate": {
          "validation": {
            "period": "MEDIUM"
          }
        },
        "certificate_validation_type": "OV",
        "tls_version": "TLS1_3",
        "cipher_reuse": "NO",
        "wildcard_certificate": "YES"
      },
      "throttle": {
        "request_size": 0,
        "response_size": 0
      },
      "timezone": "GMT",
      "nss": {
        "service": {
          "ip": "192.168.0.100"
        }
      },
      "total": {
        "size": 250279
      },
      "file": {
        "type": "Jpeg Files",
        "class": "Images"
      },
      "bandwidth_class_name": "General Surfing",
      "content_type": "image/jpeg",
      "hour": 6,
      "client": {
        "public_ip": "4.3.2.1",
        "tls_version": "TLS1_3",
        "ssl": {
          "fail_reason": "N/A",
          "fail_count": 0
        },
        "cipher_reuse": "NO"
      },
      "department": "Somewhere",
      "user_agent": {
        "token": "Microsoft Chromium Edge (127.x)"
      },
      "policy": {
        "reason": "INSPECTED"
      },
      "epochtime": "2024-07-31T06:54:37.000Z",
      "product_version": "6.2r.620_prod.84.400964_114",
      "bandwidth_throttle": "NO",
      "month_of_year": 7,
      "dlp": {
        "identifier": "0"
      },
      "datacenter": {
        "country": "XX",
        "city": "Xx",
        "name": "XXXX"
      },
      "url": {
        "category_method": "Database A",
        "category": {
          "super": "Internet Communication",
          "sub": "Webmail"
        },
        "class": "Business Use"
      },
      "minute": 54,
      "bandwidth_rule_name": "No Bandwidth Control",
      "response": {
        "code": "200",
        "payload": 245351,
        "header_size": 353
      },
      "is_ssl_certificate_expired": "NO",
      "location": "SOME_LOCATION",
      "risk": {
        "score": 0
      }
    }
  },
  "log": {
    "file": {
      "path": "REDACTED"
    },
    "offset": 0
  },
  "destination": {
    "domain": "mail.163.com",
    "ip": "1.2.3.4"
  },
  "source": {
    "nat": {
      "ip": "4.3.2.1"
    },
    "port": 35710,
    "ip": "4.3.2.1"
  },
  "url": {
    "path": "n/a://mail.163.com/js6/s",
    "original": "n/a://mail.163.com/js6/s?func=mbox:getMessageData&mid=REDACTED",
    "query": "func=mbox:getMessageData&mid=REDACTED"
  },
  "tags": ["forwarded", "zscaler_zia-web"],
  "network": {
    "protocol": "n/a"
  },
  "cloud": {
    "provider": "zscalertwo.net",
    "region": "eu-west-1",
    "account": {
      "id": "287850731128"
    }
  },
  "@timestamp": "2024-07-31T06:54:37.000Z",
  "file": {
    "extension": ["jpeg"],
    "name": ["Image.jpg"],
    "type": "file"
  },
  "ecs": {
    "version": "8.11.0"
  },
  "related": {
    "ip": ["4.3.2.1", "0.0.0.0", "192.168.0.100", "1.2.3.4"],
    "user": ["user", "user@company.com"]
  },
  "data_stream": {
    "namespace": "default",
    "type": "logs",
    "dataset": "zscaler_zia.web"
  },
  "organization": {
    "name": "COMPANY"
  },
  "http": {
    "request": {
      "referrer": "mail.163.com/js6/read/readhtml.jsp?mid=REDACTED",
      "method": "GET",
      "bytes": 4575
    },
    "response": {
      "bytes": 245704
    },
    "version": "1.1"
  },
  "tls": {
    "cipher": "TLS1_3_CK_AES_256_GCM_SHA384"
  },
  "event": {
    "agent_id_status": "missing",
    "reason": "Allowed",
    "ingested": "2024-07-31T07:26:51Z",
    "kind": "event",
    "timezone": "GMT",
    "action": "allowed",
    "id": "999999999999",
    "category": ["web"],
    "type": ["info"],
    "dataset": "zscaler_zia.web"
  },
  "user": {
    "domain": "company.com",
    "name": "user",
    "email": "user@company.com"
  },
  "user_agent": {
    "original": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0",
    "os": {
      "name": "Windows",
      "version": "10",
      "full": "Windows 10"
    },
    "name": "Edge",
    "device": {
      "name": "Other"
    },
    "version": "127.0.0.0"
  }
}
```

### What did you expect to see?

I expect the URL to be correctly set up with the expected scheme:

```json
{
  "zscaler_zia": {
    "web": {
      "referer": {
        "host": "mail.163.com"
      },
      "year": 2024,
      "alpn_protocol": "N/A",
      "day_of_month": 31,
      "second": 37,
      "user_location_name": "Unknown",
      "day": "Wed",
      "obfuscated": {
        "bendwidth": {
          "class_name": "General Surfing"
        },
        "client": {
          "public": {
            "ip": "428067240"
          },
          "ip": "428067240"
        },
        "login": "2195246850",
        "url": {
          "category": "Webmail"
        }
      },
      "ssl_decrypted": "Yes",
      "app": {
        "risk_score": "3",
        "name": "163 Mail",
        "class": "Webmail"
      },
      "module": "Webmail",
      "traffic_redirect_method": "PAC",
      "is_ssl_certificate_selfsigned": "NO",
      "forward_gateway": {
        "ip": "0.0.0.0"
      },
      "flow_type": "ZIA",
      "month": "Jul",
      "eedone": "No",
      "bypassed": {
        "time": "1970-01-01T00:00:00.000Z",
        "traffic": "0"
      },
      "key_protection_type": "Software Protection",
      "is_ssl_certificate_untrusted": "No",
      "request": {
        "payload": 0,
        "header_size": 4575
      },
      "server": {
        "cipher": "TLS1_3_CK_AES_256_GCM_SHA384",
        "certificate_validation_chain": "PASS",
        "certificate": {
          "validation": {
            "period": "MEDIUM"
          }
        },
        "certificate_validation_type": "OV",
        "tls_version": "TLS1_3",
        "cipher_reuse": "NO",
        "wildcard_certificate": "YES"
      },
      "throttle": {
        "request_size": 0,
        "response_size": 0
      },
      "timezone": "GMT",
      "nss": {
        "service": {
          "ip": "192.168.0.100"
        }
      },
      "total": {
        "size": 250279
      },
      "file": {
        "type": "Jpeg Files",
        "class": "Images"
      },
      "bandwidth_class_name": "General Surfing",
      "content_type": "image/jpeg",
      "hour": 6,
      "client": {
        "public_ip": "4.3.2.1",
        "tls_version": "TLS1_3",
        "ssl": {
          "fail_reason": "N/A",
          "fail_count": 0
        },
        "cipher_reuse": "NO"
      },
      "department": "Somewhere",
      "user_agent": {
        "token": "Microsoft Chromium Edge (127.x)"
      },
      "policy": {
        "reason": "INSPECTED"
      },
      "epochtime": "2024-07-31T06:54:37.000Z",
      "product_version": "6.2r.620_prod.84.400964_114",
      "bandwidth_throttle": "NO",
      "month_of_year": 7,
      "dlp": {
        "identifier": "0"
      },
      "datacenter": {
        "country": "XX",
        "city": "Xx",
        "name": "XXXX"
      },
      "url": {
        "category_method": "Database A",
        "category": {
          "super": "Internet Communication",
          "sub": "Webmail"
        },
        "class": "Business Use"
      },
      "minute": 54,
      "bandwidth_rule_name": "No Bandwidth Control",
      "response": {
        "code": "200",
        "payload": 245351,
        "header_size": 353
      },
      "is_ssl_certificate_expired": "NO",
      "location": "SOME_LOCATION",
      "risk": {
        "score": 0
      }
    }
  },
  "log": {
    "file": {
      "path": "REDACTED"
    },
    "offset": 0
  },
  "destination": {
    "domain": "mail.163.com",
    "ip": "1.2.3.4"
  },
  "source": {
    "nat": {
      "ip": "4.3.2.1"
    },
    "port": 35710,
    "ip": "4.3.2.1"
  },
  "url": {
    "path": "https://mail.163.com/js6/s",
    "original": "https://mail.163.com/js6/s?func=mbox:getMessageData&mid=REDACTED",
    "query": "func=mbox:getMessageData&mid=REDACTED"
  },
  "tags": ["forwarded", "zscaler_zia-web"],
  "network": {
    "protocol": "n/a"
  },
  "cloud": {
    "provider": "zscalertwo.net",
    "region": "eu-west-1",
    "account": {
      "id": "287850731128"
    }
  },
  "@timestamp": "2024-07-31T06:54:37.000Z",
  "file": {
    "extension": ["jpeg"],
    "name": ["Image.jpg"],
    "type": "file"
  },
  "ecs": {
    "version": "8.11.0"
  },
  "related": {
    "ip": ["4.3.2.1", "0.0.0.0", "192.168.0.100", "1.2.3.4"],
    "user": ["user", "user@company.com"]
  },
  "data_stream": {
    "namespace": "default",
    "type": "logs",
    "dataset": "zscaler_zia.web"
  },
  "organization": {
    "name": "COMPANY"
  },
  "http": {
    "request": {
      "referrer": "mail.163.com/js6/read/readhtml.jsp?mid=REDACTED",
      "method": "GET",
      "bytes": 4575
    },
    "response": {
      "bytes": 245704
    },
    "version": "1.1"
  },
  "tls": {
    "cipher": "TLS1_3_CK_AES_256_GCM_SHA384"
  },
  "event": {
    "agent_id_status": "missing",
    "reason": "Allowed",
    "ingested": "2024-07-31T07:26:51Z",
    "kind": "event",
    "timezone": "GMT",
    "action": "allowed",
    "id": "999999999999",
    "category": ["web"],
    "type": ["info"],
    "dataset": "zscaler_zia.web"
  },
  "user": {
    "domain": "company.com",
    "name": "user",
    "email": "user@company.com"
  },
  "user_agent": {
    "original": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0",
    "os": {
      "name": "Windows",
      "version": "10",
      "full": "Windows 10"
    },
    "name": "Edge",
    "device": {
      "name": "Other"
    },
    "version": "127.0.0.0"
  }
}
```

### Anything else?

As far as my understanding of the pipeline goes, a [rename processor](https://github.com/elastic/integrations/blob/main/packages/zscaler_zia/data_stream/web/elasticsearch/ingest_pipeline/default.yml#L855) move the `json.proto` field into `zscaler_zia.web.prototype` which is used later in a [script processor](https://github.com/elastic/integrations/blob/main/packages/zscaler_zia/data_stream/web/elasticsearch/ingest_pipeline/default.yml#L867) to populate the `network.protocol` if it hasn't already been set. However, before that, a [set processor](https://github.com/elastic/integrations/blob/main/packages/zscaler_zia/data_stream/web/elasticsearch/ingest_pipeline/default.yml#L861) uses the value from `zscaler_zia.web.alpn_protocol` to populate `network.protocol`. In all my cases, the value of `zscaler_zia.web.alpn_protocol` is `N/A` which causes the `network.protocol` to be `n/a` as well and thus generates a wrong URL starting with `n/a://......` instead of using `json.proto`-based value which is `HTTPS` (and would be properly handled by the script processor). That badly-formatted URL can not then be processed properly as its scheme won't be identified.

Would it be possible to fix the protocol handling either by inverting the execution order of the json.proto rename processor and the alpn_protocol set processor or by removing the `zscaler_zia.web.alpn_protocol` field if its value is `N/A`?

**Comment by efd6** [comment]
@srilumpa I think this will be fixed by #10657. I have tried the pipeline there with the `event.original` that you provided, with the addition that the documentation in that change requires and I see a correctly formed `url` object.
```
            "url": {
                "domain": "mail.163.com",
                "full": "https://mail.163.com/js6/s?func=mbox:getMessageData&mid=REDACTED",
                "original": "https://mail.163.com/js6/s?func=mbox:getMessageData&mid=REDACTED",
                "path": "/js6/s",
                "query": "func=mbox:getMessageData&mid=REDACTED",
                "scheme": "https"
            },
```
@kcreddy It may be worthwhile adding the event here as a test case into that change.

**Comment by kcreddy** [comment]
@srilumpa The PR  https://github.com/elastic/integrations/pull/10657 is merged and the version `3.0.3` should fix the issue.

**Comment by srilumpa** [comment]
Yes it fixed it, thank you very much.
